File: packages/typescript/src/tokens.ts
import { createRequire } from "node:module";

import type { TokenCounter } from "./types.ts";

/**
 * v0 heuristic token counter (SPEC §7): `ceil(utf16-code-units / 4)`, with
 * `""` → 0. Deterministic, dependency-free, and the shared cross-language
 * baseline every implementation MUST agree on. Real tokenizers are layered on
 * top via {@link createTokenCounter} or any user-supplied
 * {@link TokenCounter}.
 */
export function estimateTokens(s: string): number {
  if (s.length === 0) return 0;
  return Math.ceil(s.length / 4);
}

/**
 * The set of tiktoken-style BPE encodings exposed by `gpt-tokenizer`. We pin
 * this list in the public type so users get autocomplete without having to
 * import a gpt-tokenizer type that may not be installed.
 */
export type TiktokenEncoding =
  | "o200k_base"
  | "cl100k_base"
  | "p50k_base"
  | "p50k_edit"
  | "r50k_base";

export type CreateTokenCounterOptions = {
  /** BPE encoding name. Defaults to `o200k_base` (GPT-4o / GPT-5 family). */
  encoding?: TiktokenEncoding;
};

/**
 * The TS package's default real-tokenizer counter. Loads the requested
 * `gpt-tokenizer` encoding (declared in `package.json` under
 * `optionalDependencies` so installs don't fail on hosts that can't ship it)
 * and returns a {@link TokenCounter} that calls `countTokens` on it.
 *
 * Throws synchronously if `gpt-tokenizer` is not installed, so callers in
 * environments that can't ship it (restricted bundlers, no-tokenizer hosts,
 * …) can catch and fall back to the heuristic.
 *
 * Defaults to the `o200k_base` encoding per the PRD. Pass `{ encoding }` to
 * target a different model family (e.g. `cl100k_base` for GPT-3.5 / GPT-4).
 */
export function createTokenCounter(
  options: CreateTokenCounterOptions = {},
): TokenCounter {
  const encoding = options.encoding ?? "o200k_base";
  const mod = loadGptTokenizerEncoding(encoding);
  return (s: string): number => mod.countTokens(s);
}

type GptTokenizerEncoding = { countTokens: (s: string) => number };

const req = createRequire(import.meta.url);

function loadGptTokenizerEncoding(
  encoding: TiktokenEncoding,
): GptTokenizerEncoding {
  try {
    return req(`gpt-tokenizer/encoding/${encoding}`) as GptTokenizerEncoding;
  } catch (cause) {
    throw new Error(
      "createTokenCounter() requires the optional 'gpt-tokenizer' dependency. " +
        "Install it with `npm install gpt-tokenizer`, or pass a custom " +
        "tokenCounter to compress().",
      { cause },
    );
  }
}

File: packages/python/sheet_compressor/tokens.py
"""Token counting (SPEC §7).

Default heuristic: ceil(utf-16 code units / 4). Optional real-tokenizer factory
using `tiktoken` (declared in `pyproject.toml` under
`project.optional-dependencies.tokenizer`).
"""

import math
from typing import Callable

TokenCounter = Callable[[str], int]


def _utf16_code_units(s: str) -> int:
    """Length of `s` in UTF-16 code units.

    Python's `len(s)` returns code points; a non-BMP character such as
    "\U0001F600" is one code point but two UTF-16 code units. The SPEC §7
    heuristic is defined in UTF-16 units so all language ports agree.
    """
    n = 0
    for ch in s:
        n += 2 if ord(ch) > 0xFFFF else 1
    return n


def estimate_tokens(s: str) -> int:
    """SPEC §7 heuristic: `ceil(utf16-code-units / 4)`, `""` -> 0."""
    if s == "":
        return 0
    return math.ceil(_utf16_code_units(s) / 4)


def create_token_counter(encoding: str = "o200k_base") -> TokenCounter:
    """Return a `TokenCounter` backed by `tiktoken`.

    Defaults to `o200k_base` (GPT-4o / GPT-5 family). Raises a clear error if
    `tiktoken` is not installed.
    """
    try:
        import tiktoken  # type: ignore[import-not-found]
    except ImportError as cause:
        raise ImportError(
            "create_token_counter() requires the optional 'tiktoken' "
            "dependency. Install it with `pip install sheet-compressor[tokenizer]` "
            "or `pip install tiktoken`, or pass a custom token_counter to compress()."
        ) from cause

    enc = tiktoken.get_encoding(encoding)

    def count(s: str) -> int:
        return len(enc.encode(s))

    return count
